proc with inline pragma passed to generic proc leads to bad codegen

```Nimrod
proc cmpx(d: int): bool {.inline.} = d > 0

proc abc[C](cx: C, d: int) =
  echo cx(d)
  
abc(cmpx, 10)
```
```text
CC: inlineprocvar
F:\projects\bug\nimcache\inlineprocvar.c:15:32: error: unknown type name 'TY_GP9aNpGnrI0ICDZLuW4Pd7A'
 typedef N_INLINE_PTR(NIM_BOOL, TY_GP9aNpGnrI0ICDZLuW4Pd7A) (NI d);
                                ^
F:\projects\bug\nimcache\inlineprocvar.c:28:46: error: unknown type name 'TY_GP9aNpGnrI0ICDZLuW4Pd7A'
 N_NIMCALL(void, abc_UlsM8cz9cJDzJMrkcr2NDWw)(TY_GP9aNpGnrI0ICDZLuW4Pd7A cx, NI d);
                                              ^
F:\projects\bug\nimcache\inlineprocvar.c:75:46: error: unknown type name 'TY_GP9aNpGnrI0ICDZLuW4Pd7A'
 N_NIMCALL(void, abc_UlsM8cz9cJDzJMrkcr2NDWw)(TY_GP9aNpGnrI0ICDZLuW4Pd7A cx, NI d) {
                                              ^
```

I think the Nim compiler should generate error message rather than let C compiler do that

more minimal case (+ case showing it only trips on proc, not template), as reported in duplicate bug https://github.com/nim-lang/Nim/issues/8512

```nim
proc fun0()  {.inline.} = discard

when defined(case_ok):
  template callFun[Fun](fun: Fun) = fun()
when defined(case_bug):
  proc callFun[Fun](fun: Fun) = fun()

callFun(fun0)
```

furthermore, the error msg was lacking 1st part, shown here:

```
rnim -d:case_bug bugs/ICE/t01_function_cannot_return_funciton_type.nim
nim c --nimcache:/tmp/nim//nimcache/ -o:/tmp/nim//app -r -d:case_bug bugs/ICE/t01_function_cannot_return_funciton_type.nim
Error: execution of an external compiler program 'clang -c  -w  -I/Users/timothee/git_clone/nim/Nim/lib -o /tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c.o /tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c' failed with exit code: 1

/tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c:24:21: error: function cannot return function type 'int (void)'
typedef N_INLINE_PTR(void, tyProc_1dK9cRbNmyUl29auLED01heQ) (void);
                    ^
/tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c:24:21: error: 'void' must be the first and only parameter if specified
/tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c:25:63: error: unknown type name 'tyProc_1dK9cRbNmyUl29auLED01heQ'
N_LIB_PRIVATE N_NIMCALL(void, callFun_IUIW1cKd2qUszJFnqOROJw)(tyProc_1dK9cRbNmyUl29auLED01heQ fun);
                                                              ^
/tmp/nim//nimcache/timn_t01_function_cannot_return_funciton_type.c:65:63: error: unknown type name 'tyProc_1dK9cRbNmyUl29auLED01heQ'
N_LIB_PRIVATE N_NIMCALL(void, callFun_IUIW1cKd2qUszJFnqOROJw)(tyProc_1dK9cRbNmyUl29auLED01heQ fun) {
                                                              ^
4 errors generated.
```

--- a/lib/nimbase.h
+++ b/lib/nimbase.h
@@ -107,6 +107,8 @@ __clang__
 #  define N_INLINE(rettype, name) rettype __inline name
 #endif
 
+#define N_INLINE_PTR(rettype, name) rettype (*name)
+
 #if defined(__POCC__)
 #  define NIM_CONST /* PCC is really picky with const modifiers */
 #  undef _MSC_VER /* Yeah, right PCC defines _MSC_VER even if it is

--- a/tests/ccgbugs/t5345.nim
+++ b/tests/ccgbugs/t5345.nim
@@ -1,0 +1,10 @@
+discard """
+  output: true
+"""
+
+proc cmpx(d: int): bool {.inline.} = d > 0
+
+proc abc[C](cx: C, d: int) =
+  echo cx(d)
+  
+abc(cmpx, 10)

--- a/tests/ccgbugs/t5701.nim
+++ b/tests/ccgbugs/t5701.nim
@@ -1,0 +1,17 @@
+discard """
+  output: '''(Field0: 1, Field1: 1)
+(Field0: 2, Field1: 2)
+(Field0: 3, Field1: 3)
+'''
+"""
+
+iterator zip[T1, T2](a: openarray[T1], b: openarray[T2]): iterator() {.inline.} =
+  let len = min(a.len, b.len)
+  for i in 0..<len:
+    echo (a[i], b[i])
+
+proc foo(args: varargs[int]) =
+  for i in zip(args,args):
+    discard
+
+foo(1,2,3)

